internal/flowcontrol/writebuffer.go

```diff
@@ -99,31 +104,43 @@ func (w *ResourceSliceWriteBuffer) processQueueItem(ctx context.Context) bool {
 	ctx = logr.NewContext(ctx, logger)
 
 	w.mut.Lock()
+	insertionTime := w.insertionTime[sliceNSN]
 	updates := w.state[sliceNSN]
 	delete(w.state, sliceNSN)
+
+	// Limit the number of operations per patch request
+	const max = (10000 / 2) - 2 // 2 ops to initialize status + 2 ops per resource
+	if len(updates) > max {
+		w.state[sliceNSN] = updates[max:]
+		updates = updates[:max]
+	} else {
+		delete(w.insertionTime, sliceNSN)
+	}
 	w.mut.Unlock()
 
+	// We only forget the rate limit once the update queue for this slice is empty.
+	// So the first write is fast, but a steady stream of writes will be throttled exponentially.
```

Comment:
For this to be true, shouldn't we always call AddRateLimited in PatchStatusAsync?

Reply:
That would back off once per call to `PatchStatusAsync` which seems too aggressive. Right now we only back off once per batch i.e. call to `updateSlice`.